### PR TITLE
Ensure endpoint names are unique to avoid failures later in execution.

### PR DESCRIPTION
**Pull request by mitchdenny**
Builds upon changes in #4064 

Moves the uniqueness check until after the endpoint is constructed and name is derived (to share the various bits of fallback logic).

Added a test case to cover this.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4071)

### CODE REVIEW

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -439,6 +434,11 @@ private static void ApplyEndpoints<T>(this IResourceBuilder<T> builder, IResourc
             isExternal: isExternal,
             isProxied: isProxied);
 
+        if (builder.Resource.Annotations.OfType<EndpointAnnotation>().Any(sb => string.Equals(sb.Name, annotation.Name, StringComparisons.EndpointAnnotationName)))
+        {
+            throw new DistributedApplicationException($"Endpoint with name '{annotation.Name}' already exists");
```
Comment (davidfowl) [inline src/Aspire.Hosting/ResourceBuilderExtensions.cs:439]: I think we should improve the error message. We default the scheme to the name so they may not even know the endpoint has a name. We should try out best to tell them that it does and that the if no name is specified, the scheme will be used as the name. Then we should recommend that they specify a unique name per endpoint.

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs [hunk 2]
```diff
@@ -439,6 +434,11 @@ private static void ApplyEndpoints<T>(this IResourceBuilder<T> builder, IResourc
             isExternal: isExternal,
             isProxied: isProxied);
 
+        if (builder.Resource.Annotations.OfType<EndpointAnnotation>().Any(sb => string.Equals(sb.Name, annotation.Name, StringComparisons.EndpointAnnotationName)))
+        {
+            throw new DistributedApplicationException($"Endpoint with name '{annotation.Name}' already exists. Endpoint name may not have been explicitly specified and was derived automatically from scheme argument (e.g. 'http', 'https', or 'tcp;). Multiple calls to WithEndpoint (and related methods) may result in a conflict if name argument is not specified. Each endpoint must have a unique name. For more information on networking in .NET Aspire see: https://aka.ms/dotnet/aspire/networking");
```
Comment (timheuer) [inline src/Aspire.Hosting/ResourceBuilderExtensions.cs:439]: typo: 'tcp; -- i assume to be 'tcp'

File: tests/Aspire.Hosting.Tests/WithEndpointTests.cs [hunk 3]
```diff
@@ -94,7 +94,22 @@ public void EndpointsWithTwoPortsSameNameThrows()
                     .WithHttpsEndpoint(3000, 2000, name: "mybinding");
         });
 
-        Assert.Equal("Endpoint with name 'mybinding' already exists", ex.Message);
+        Assert.Equal("Endpoint with name 'mybinding' already exists. Endpoint name may not have been explicitly specified and was derived automatically from scheme argument (e.g. 'http', 'https', or 'tcp;). Multiple calls to WithEndpoint (and related methods) may result in a conflict if name argument is not specified. Each endpoint must have a unique name. For more information on networking in .NET Aspire see: https://aka.ms/dotnet/aspire/networking", ex.Message);
```
Comment (timheuer) [inline tests/Aspire.Hosting.Tests/WithEndpointTests.cs:97]: Similar tcp; typos in the asserts.

**Review by timheuer — CHANGES_REQUESTED** [review]
typos

**Comment by mitchdenny** [comment]
@timheuer thanks for spotting those!

**Comment by mitchdenny** [comment]
@timheuer can you reset your vote?

**Review by timheuer — APPROVED** [review]

**Pull request merged**